src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1628,12 +1628,16 @@ private void HandleExecuteSubmissionException(GraphBuildSubmission submission, E
                 ? ae.InnerExceptions.First()
                 : ex;
 
-            if (submission.IsStarted)
+            lock (_syncLock)
             {
-                submission.CompleteResults(new GraphBuildResult(submission.SubmissionId, ex));
+                if (submission.IsStarted)
+                {
+                    submission.CompleteResults(new GraphBuildResult(submission.SubmissionId, ex));
+                }
+
+                _overallBuildSuccess = false;
```

Comment:
There's a similar write at:

https://github.com/dotnet/msbuild/blob/b02f8de7e24d541df1d8c7d71cf8036d0ce91213/src/Build/BackEnd/BuildManager/BuildManager.cs#L1608

Why is it not protected by the lock as well?

Reply:
On line 857, 2403, and 2417. Though when they're called there might have been locked already.

Reply:
I tried to add the minimal number of locks (though I had missed the one @ladipro mentioned). If a write is not under a lock it means the method is called from other methods that do lock (or I missed them).  Do you prefer if I just lock every single write regardless if it's locked in a calling method?

Reply:
> Do you prefer if I just lock every single write regardless if it's locked in a calling method?
definitely not 🙂 I did a quick crtl+f and commented while realizing some may have already been locked

Reply:
As a general practice, you might consider adding `Debug.Assert(Monitor.IsEntered(lockobj))` at the top of all methods that need to be called within a lock, so it's not necessary to mentally infer it.

Reply:
I've done this in other projects and it's been very handy to protect against future maintenance errors

Reply:
Well that's a cool trick to keep around! I added checks to non-locking but state mutating methods in the BuildManager and found a few that weren't protected.